### PR TITLE
Candles HLOC Data

### PR DESCRIPTION
**Pull request by adamfraser**
Changelist

- Add two fields to the to Candles DB object (orderbookMidPriceOpen & orderbookMidPriceClose)
- Add function to get the mid price from the orderbook cache
- Track the orderbookMidPriceOpen & Close at the start and end of each candle period
- Return the new fields in the API and Websocket responses

Test Plan
Tested manually through WS requests and API 

## Summary by CodeRabbit

- **New Features**
	- Introduced properties for `orderbookMidPriceOpen` and `orderbookMidPriceClose` in the candle data structure, enhancing market data tracking.
	- Added functionality to the candle generation tests to validate order book mid-price calculations and updates.
	- Implemented a migration script to add mid-price columns to the candles table in the database.
	- Enhanced the testing suite for the orderbookLevelsCache module to include mid-price calculation validation.
	- Introduced a new function to calculate the mid-price of an order book, improving data retrieval for users.

- **Bug Fixes**
	- Improved test coverage for candle generation logic related to order book interactions, ensuring accurate price handling.

- **Documentation**
	- Updated API documentation to reflect new mid-price fields in responses, enhancing market condition insights for users.
	- Modified Swagger documentation to include new fields for mid-prices, ensuring clarity for API consumers.

### CODE REVIEW

File: .github/workflows/indexer-build-and-push-dev-staging.yml [hunk 1]
```diff
@@ -6,53 +6,14 @@ on: # yamllint disable-line rule:truthy
       - main
       - 'release/indexer/v[0-9]+.[0-9]+.x'  # e.g. release/indexer/v0.1.x
       - 'release/indexer/v[0-9]+.x'  # e.g. release/indexer/v1.x
+      - 'adam/ct-973-add-midbookopen-and-midbookclose-values-to-candles'
```
Comment (dydxwill) [inline .github/workflows/indexer-build-and-push-dev-staging.yml:9]: revert changes in this file

File: indexer/packages/postgres/src/models/candle-model.ts [hunk 2]
```diff
@@ -77,4 +77,8 @@ export default class CandleModel extends Model {
   trades!: number;
 
   startingOpenInterest!: string;
+
+  orderbookMidPriceOpen: string | undefined;
```
Comment (dydxwill) [inline indexer/packages/postgres/src/models/candle-model.ts:81]: orderbookMidPriceOpen?: string;

File: indexer/packages/redis/__tests__/caches/orderbook-levels-cache.test.ts [hunk 3]
```diff
@@ -685,4 +686,93 @@ describe('orderbookLevelsCache', () => {
...
+      ]);
+
+      const midPrice = await getOrderBookMidPrice(ticker, client);
+      expect(midPrice).toEqual(45350);
+    });
+
+  });
+
+  it('returns undefined if there are no bids or asks', async () => {
+    await Promise.all([updatePriceLevel({
```
Comment (dydxwill) [inline indexer/packages/redis/__tests__/caches/orderbook-levels-cache.test.ts:753]: no need for promise.all here

can just do `await updatePriceLevel`

File: indexer/packages/redis/__tests__/caches/orderbook-levels-cache.test.ts [hunk 4]
```diff
@@ -685,4 +686,93 @@ describe('orderbookLevelsCache', () => {
...
+      sizeDeltaInQuantums: '2000',
+      client,
+    })]);
+
+    const midPrice = await getOrderBookMidPrice(ticker, client);
+    expect(midPrice).toBeUndefined();
+  });
+
+  it('returns undefined if humanPrice is NaN', async () => {
+    await Promise.all([updatePriceLevel({
```
Comment (dydxwill) [inline indexer/packages/redis/__tests__/caches/orderbook-levels-cache.test.ts:766]: no need for promise.all here

can just do `await updatePriceLevel`

File: indexer/services/ender/__tests__/helpers/redis-helpers.ts [hunk 5]
```diff
@@ -29,3 +32,34 @@ export async function expectStateFilledQuantums(
   expect(stateFilledQuantums).toBeDefined();
   expect(stateFilledQuantums).toEqual(quantums);
 }
+
+export async function updatePriceLevel(
```
Comment (dydxwill) [inline indexer/services/ender/__tests__/helpers/redis-helpers.ts:36]: This helper function doesn't seem necessary

File: indexer/services/ender/__tests__/helpers/redis-helpers.ts [hunk 6]
```diff
@@ -29,3 +32,34 @@ export async function expectStateFilledQuantums(
...
+  await OrderbookLevelsCache.updatePriceLevel({
+    ticker,
+    side,
+    humanPrice: price,
+    sizeDeltaInQuantums: quantums,
+    client: redisClient,
+  });
+}
+
+export function clearOrderbookLevelsCacheForTests() {
```
Comment (dydxwill) [inline indexer/services/ender/__tests__/helpers/redis-helpers.ts:52]: is this needed? we already have `await redis.deleteAllAsync(redisClient);`

File: indexer/services/ender/src/lib/candles-generator.ts [hunk 7]
```diff
@@ -432,6 +466,28 @@ export class CandlesGenerator {
...
+
+    const candle: CandleUpdateObject = {
+      id: existingCandle.id,
+      low: existingCandle.low,
+      high: existingCandle.high,
+      close: existingCandle.close,
+      baseTokenVolume: existingCandle.baseTokenVolume,
+      usdVolume: existingCandle.usdVolume,
+      trades: existingCandle.trades,
+      orderbookMidPriceOpen: existingCandle.orderbookMidPriceOpen ?? undefined,
```
Comment (dydxwill) [inline indexer/services/ender/src/lib/candles-generator.ts:489]: Don't think you need this line. If orderbookMidPriceOpen is not included in the update object, it won't be updated.

File: indexer/packages/postgres/src/models/candle-model.ts (hunk 2 above)
Comment (adamfraser) [reply indexer/packages/postgres/src/models/candle-model.ts:81]: I think having the property be required but allowing the type to be undefined is easier to understand and safer in code.

File: indexer/services/ender/src/lib/candles-generator.ts (hunk 7 above)
Comment (adamfraser) [reply indexer/services/ender/src/lib/candles-generator.ts:489]: The `existingCandle` is a `CandleFromDatabase` type which can be null so we need to coerce that into undefined.

**Review by dydxwill — APPROVED** [review]

**Comment by adamfraser** [comment]
@mergifyio backport release/indexer/v5.x

**Pull request merged**